### ISSUE TITLE
Pre-install certain JupyterLab extensions

### ISSUE THREAD

**Issue by yannickperrenet** [orchest/orchest#169]
The idea is to ship JupyterLab (through the `jupyter-server`) with a set of pre-installed extensions.

Ideally, the extensions can be considered:

* (reasonably) mature,
* used by many users (or wanted but hard to set up),
* lightweight, in the sense that they do not completely bloat the Docker image.

Feel free to add any extensions here that you would like to see pre-installed. A few initial ideas:

* [jupyter-lsp](https://github.com/krassowski/jupyterlab-lsp): LSP support
* [xeus-python](https://github.com/jupyter-xeus/xeus-python): Different kernel to allow for a debugger. This is considered more than just an extension though.

**Comment by ricklamers** [comment]
Love this idea!

**Comment by yannickperrenet** [comment]
* [jupyterlab-git](https://github.com/jupyterlab/jupyterlab-git): Extension for version control using git.
* [jupyterlab-variableInspector](https://github.com/lckr/jupyterlab-variableInspector): Shows the currently used variables and their values.
* [toc2](https://github.com/ipython-contrib/jupyter_contrib_nbextensions/tree/master/src/jupyter_contrib_nbextensions/nbextensions/toc2): Table of content.

**Comment by ricklamers** [comment]
Great suggestions!

> toc2: Table of content.

This is already built in to JupyterLab (at least in 3.0.6)

**Comment by yannickperrenet** [comment]
@ricklamers Has done some work on this in https://github.com/orchest/orchest/commit/1e7843434480cb1a3207abe2e66ad393185148f3 pre-installing the `jupyter-git` extension.

**Comment by yannickperrenet** [comment]
I came across another extension that we could add:
* https://github.com/nbsafety-project/nbsafety

**Comment by andtheWings** [comment]
May I ask about the status of this improvement? I see that the jupyterlab-git extension is installed, but I would love to have jupyter-lsp as well. 

I tried putting `pip install 'jupyterlab>=3.0.0,<4.0.0a0' jupyterlab-lsp` in my main build script and `install.packages("languageserver", dependencies = TRUE, repos="http://cran.rstudio.com/")` in my install.R file, but it doesn't seem to have worked on restarting the jupyterlab environment. What would be the correct method for adding my user-specific extensions?

**Comment by yannickperrenet** [comment]
Good question. At the time we decided not to pre-install a bunch of extension, since that would bloat the `jupyter-server` image everyone has to pull/install. Instead we chose to make it possible to install extensions by allowing users themselves to configure JupyterLab.

> What would be the correct method for adding my user-specific extensions?

1. Settings > Configure JupyterLab
2. The script you see here works similarly as the environment build, but instead build a new JupyterLab image. So here you can enter your `pip install`
3. Press Build
4. Make sure the session is restarted since that will reboot your JupyterLab. (This should happen automatically.)

I hope that answers your question.

**Comment by andtheWings** [comment]
Thanks for the guidance! I got `pip install 'jupyterlab>=3.0.0,<4.0.0a0' jupyterlab-lsp` installed with in the jupyterlab cofig environment. Now I just need the R-implementation of the server installed. I tried `R -e 'install.packages("languageserver")` and `conda install -c conda-forge r-languageserver`, but it said the R and conda commands were not found. Any further advice?

**Comment by yannickperrenet** [comment]
Sure. Both `conda` and `R` are not installed by default and so to install the language server you first need to install them.

For `conda` you can do this with:
```sh
mkdir -p ~/miniconda3

# jupyter-server runs on Python3.8
apt-get install -y wget
wget https://repo.anaconda.com/miniconda/Miniconda3-py38_4.10.3-Linux-x86_64.sh -O ~/miniconda3/miniconda.sh

# -b: Install unattended
bash ~/miniconda3/miniconda.sh -b -u

# Refresh terminal without needing to close it
export PATH="/root/miniconda3/bin:$PATH"
source ~/.bashrc

conda install -c conda-forge r-languageserver
```

I am not really familiar with the R ecosystem so I am afraid I can't help you past `apt-get install -y r-base`.

**Comment by andtheWings** [comment]
Sorry, this ended up being way more complicated than I intended. When I try running the script, I get `E: Unable to locate package wget`, which seems weird because it looks like the container is using debian bullseye repository, which has wget.

**Comment by yannickperrenet** [comment]
Yes, there are definitely things there to improve. If you have any ideas, would love to hear them.

Regarding the `wget`, that is why I added the `apt-get install -y wget` line. 

The `jupyter-server` container uses `python:3.8-slim` as a base image and so certain things would indeed be missing.

**Comment by andtheWings** [comment]
Yes, that's the error I get when I enter `apt-get install -y wget`, the log returns `E: Unable to locate package wget`

**Comment by yannickperrenet** [comment]
That is strange, since we would be running the code inside the same Docker image (unless you are on a very old version of Orchest).

In general to fix the `E: Unable to locate package wget` issue you can first run `apt-get update` before `apt-get install -y wget`

**Comment by astrojuanlu** [comment]
> What I want to avoid is us having to make sure all extensions are installed correctly (what about maintainability? what if we later decide to remove a certain extension?) and do not cause any conflicts with other extensions a user might want to install.

Yep, that's a valid point as well. However, I'm thinking if Orchest could provide value by preinstalling some commonly used extensions, so that the maintainability cost is justified.

> If someone doesn't know extensions exists, then they probably haven't looked for them (maybe because they don't care).

That's a reasonable explanation - but another one is that they don't know they want them 😃 

> I think it makes sense to be closer to vscode than something like vim. But in the end vscode doesn't ship with all available extensions pre-installed.

Speaking of which, nowadays I tend to use a vim distribution that installs a bunch of stuff 🤓 https://vim.fisadev.com/ The maintainer has to make sure that everything works, but the added value is enormous.

> What extensions do you suggest we add?

I remember that https://jupyter-contrib-nbextensions.readthedocs.io/ provides a nice UI to pick some popular ones, but (emphasis mine)

> Most nbextensions here should work fine with jupyterhub (because jupyterhub spawns regular notebook servers for each individual user), **but won’t work with jupyterlab** (because the jupyterlab javascript framework is different to notebook’s, and still rapidly changing under active development).

So, with JLab we're kind of back to square 1 I fear. A bit of research is needed to decide what could be preinstalled - but first we'd have to agree that this is a good idea, within some limits.

**Comment by yannickperrenet** [comment]
As a general rule of thumb I would be fine to include an extension, if:
* It can be considered stable. It is better to have a working IDE that is missing some features than a breaking IDE with some cool features.
* It can be agreed upon that the extension is useful for 80% of all users. I would say lsp support is such an extension.

**Comment by yannickperrenet** [comment]
@astrojuanlu Feel free to reopen the issue if you are keen on including additional extensions and have some good suggestions :wink:

**Comment by astrojuanlu** [comment]
100 % with both criteria. I already have a list of extensions I would consider useful or popular, will curate it in the coming days.

I can't reopen since I don't have permissions for it 🙃

**Comment by astrojuanlu** [comment]
Related: https://github.com/jupyterlab-contrib/jupyterlab-contrib.github.io/issues/23

**Comment by astrojuanlu** [comment]
Also: https://jupyterlab-contrib.github.io/migrate_from_classical.html

**Comment by astrojuanlu** [comment]
Consolidating a few of the alternatives we mentioned here, in order of increasing perceived complexity or risk:

- [jupyterlab-execute-time](https://github.com/deshaw/jupyterlab-execute-time) (been using this one everywhere lately, it's super useful - it requires proper configuration)
- [jupyterlab_code_formatter](https://github.com/ryantam626/jupyterlab_code_formatter) (requires some configuration to work the first time, if we install this we should ship the proper config too)
- [jupyterlab-variableInspector](https://github.com/lckr/jupyterlab-variableInspector) (never used it but it's one of the most requested features of Jupyter for MATLAB, RStudio and Spyder users)
- [jupyter-lsp](https://github.com/krassowski/jupyterlab-lsp) (again, never used it but I probably should)
- [nbsafety](https://github.com/nbsafety-project/nbsafety) (although it requires a separate kernel, we should look into it carefully)

Already installed or not needed:

- xeus-python (debugging already works since #786)
- jupyterlab-git (already preinstalled, see [docs](https://docs.orchest.io/en/latest/fundamentals/projects.html#git-inside-orchest))

Just having the first one would be nice, having the first three properly configured would be very cool, and having them all would be awesome.

**Comment by yannickperrenet** [comment]
To quickly share my thoughts (on https://github.com/orchest/orchest/issues/169#issuecomment-1124903038):

| extension | thought|
| --- | --- |
| jupyterlab-execute-time | :white_check_mark: |
| jupyterlab_code_formatter | :white_check_mark: |
| jupyterlab-variableInspector | :x: with the addition of the debugger I don't think it is needed.  |
| jupyterlab-lsp | :question: would need testing on performance hit given that we need to run a language server (or multiple depending on how the plugin works, e.g. one for every open file) |

And definitely providing a sane default config that users can edit if they want to.

**Comment by astrojuanlu** [comment]
I just confirmed that the variable inspector that comes with the native Jupyter debugger covers the use cases I'd expect. In addition, [jupyterlab-variableInspector](https://github.com/lckr/jupyterlab-variableInspector) seems to have some issues and not much maintenance.

Let's start by preinstalling jupyterlab-execute-time and jupyterlab_code_formatter then, and then we can re-evaluate the rest on a case-by-case basis.

**Comment by astrojuanlu** [comment]
We should try https://github.com/jupyter-server/jupyter-resource-usage too, although I don't know if it would properly detect the RAM consumed by the dockerized kernel.

**Comment by astrojuanlu** [comment]
Update: I don't think jupyter-resource-usage or jupyterlab-system-monitor work with remote/gateway kernels ❌ and I can't get jupyterlab-kernel-usage to work either ❌ so I'm [back to my earlier comment](https://github.com/orchest/orchest/issues/169#issuecomment-1129164739): let's try jupyterlab-execute-time and jupyterlab_code_formatter.

**Comment by astrojuanlu** [comment]
`jupyterlab-execute-time` requires changing a user setting:

```json
{
    "recordTiming": true
}
```

![image](https://user-images.githubusercontent.com/316517/196362062-61f37d7f-67d6-4a3f-923d-a57045a727da.png)

any pointers on where to ship such settings in Orchest? I can try to open a pull request myself.
